pkgs/development/python-modules/bitsandbytes/default.nix

```diff
@@ -48,47 +52,51 @@ buildPythonPackage {
   inherit pname version;
   pyproject = true;
 
-  disabled = pythonOlder "3.7";
-
   src = fetchFromGitHub {
     owner = "TimDettmers";
     repo = "bitsandbytes";
     rev = "refs/tags/${version}";
-    hash = "sha256-GFbFKPdV96DXPA+PZO4h0zdBclN670fb0PGv4QPHWHU=";
+    hash = "sha256-yvxD5ymMK5p4Xg7Csx/90mPV3yxUC6QUuF/8BKO2p0k=";
   };
 
-  postPatch =
-    ''
-      substituteInPlace Makefile --replace "/usr/bin/g++" "g++" --replace "lib64" "lib"
-      substituteInPlace bitsandbytes/cuda_setup/main.py  \
-        --replace "binary_path = package_dir / self.binary_name"  \
-                  "binary_path = Path('$out/${python.sitePackages}/${pname}')/self.binary_name"
-    ''
-    + lib.optionalString torch.cudaSupport ''
-      substituteInPlace bitsandbytes/cuda_setup/main.py  \
-        --replace "/usr/local/cuda/lib64" "${cuda-native-redist}/lib"
-    '';
+  # By default, which library is loaded depends on the result of `torch.cuda.is_available()`.
+  # When `cudaSupport` is enabled, bypass this check and load the cuda library unconditionnally.
+  # Indeed, in this case, only `libbitsandbytes_cuda124.so` is built. `libbitsandbytes_cpu.so` is not.
+  # Also, hardcode the path to the previously built library instead of relying on
+  # `get_cuda_bnb_library_path(cuda_specs)` which relies on `torch.cuda` too.
+  postPatch = lib.optionalString cudaSupport ''
+    substituteInPlace bitsandbytes/cextension.py \
+      --replace-fail "if cuda_specs:" "if True:" \
+      --replace-fail \
+        "cuda_binary_path = get_cuda_bnb_library_path(cuda_specs)" \
+        "cuda_binary_path = PACKAGE_DIR / 'libbitsandbytes_cuda${cudaVersionString}.so'"
+  '';
 
   CUDA_HOME = "${cuda-native-redist}";
 
-  preBuild =
-    if torch.cudaSupport then
-      with torch.cudaPackages;
-      let
-        cudaVersion = lib.concatStrings (lib.splitVersion torch.cudaPackages.cudaMajorMinorVersion);
-      in
-      ''make CUDA_VERSION=${cudaVersion} cuda${cudaMajorVersion}x''
-    else
-      ''make CUDA_VERSION=CPU cpuonly'';
-
-  nativeBuildInputs = [
+  build-system = [
+    cmake
     setuptools
-    wheel
-  ] ++ lib.optionals torch.cudaSupport [ cuda-native-redist ];
+  ] ++ lib.optionals cudaSupport [ cuda-native-redist ];
+
+  dontUseCmakeConfigure = true;
 
-  buildInputs = lib.optionals torch.cudaSupport [ cuda-redist ];
+  buildInputs = lib.optionals cudaSupport [ cuda-redist ];
+
+  preBuild =
+    if cudaSupport then
+      ''
+        export NVCC_APPEND_FLAGS="-I${cuda-native-redist}/include -L${cuda-native-redist}/lib"
+        cmake -DCMAKE_CXX_FLAGS="-I${cuda-native-redist}/include" -DCOMPUTE_BACKEND=cuda -S .
```

Comment:
This is normally done in `confgiurePhase` which `pkgs.cmake` in `nativeBuildInputs` would set automatically. Istead of specifying the `-D` flags in here, maybe move them to the nix attrset as `cmakeFlags` so the hook passes them on as a bash variable. If for some reason you choose not to use the hook but to call cmake yourself, reimplement the `flagsArray=(...) ; concatTo ...` functionality from the hook

Reply:
I was able to successfully rely on the cmake hook.
I still manually run `make` in the `preBuild` phase.